### PR TITLE
Feature 164 add dialog for joining or creating group

### PR DESCRIPTION
**Pull request by Jdallos**

### CODE REVIEW

File: src/components/App.tsx [hunk 1]
```diff
@@ -40,6 +42,10 @@ const App: React.FC = () => {
               path="/surveys/:surveyId/*"
               element={ <SurveyScreen/> }
             />
+            <Route
+              path="/groups/:groupId*"
```
Comment (lauriLukkarinen) [inline src/components/App.tsx:46]: Should this include a slash also?

```suggestion
              path="/groups/:groupId/*"
```

File: src/components/group/group-members.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ * Component properties
+ */
+interface Props {
+  groupMembers: User[];
+}
+
+/**
+ * Component for Group Members
+ */
+const GroupMembers: FC<Props> = ({ groupMembers }) => {
```
Comment (lauriLukkarinen) [inline src/components/group/group-members.tsx:19]:
```suggestion
/**
 * Component for Group Members
 *
 * @param props component properties
 */
const GroupMembers: FC<Props> = ({ groupMembers }) => {
```

File: src/components/group/group-members.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+              style={{ marginRight: 8 }}
+              onClick={ () => console.log(params.row)}
+            >
+              { strings.groupManagementScreen.groupMembersScreen.role }
+            </Button>
+            <Button
+              startIcon={ <PersonRemoveIcon/> }
+              color="error"
+              // TODO: delete user form group logic
+              onClick={ () => console.log(params.row)}
```
Comment (lauriLukkarinen) [inline src/components/group/group-members.tsx:63]:
```suggestion
              // onClick={ ... }
```

File: src/components/group/group-members.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    return (
+      <Paper>
+        <DataGrid
+          autoHeight
+          rows={ groupMembers }
+          columns={ columns }
+          pageSize={ 10 }
+          disableSelectionOnClick
+          onRowClick={ () => {} }
```
Comment (lauriLukkarinen) [inline src/components/group/group-members.tsx:80]:
```suggestion
```

File: src/components/group/group-members.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        />
+      </Paper>
+    );
+  };
+
+  return (
+    <>
+      { renderGroupMemberDataTable() }
+    </>
+  );
```
Comment (lauriLukkarinen) [inline src/components/group/group-members.tsx:90]: No need to create this additional rendering layer, just render the table directly in the component return structure and move the definition of the columns also directly under component.

You can even go a bit further, and wrap the column structure with a useMemo(), so it only updates when necessary. This is not required, though. Then you need to use the language from Redux as a dependency, so the structure actually updates when the language updates.

File: src/components/group/pending-invites.tsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+ * Component properties
+ */
+interface Props {
+  pendingInvites: GroupJoinInvite[];
+}
+
+/**
+ * Component for pending invites
+ */
+const PendingInvites: FC<Props> = ({ pendingInvites }) => {
```
Comment (lauriLukkarinen) [inline src/components/group/pending-invites.tsx:18]:
```suggestion
/**
 * Component for pending invites
 *
 * @param props component properties
 */
const PendingInvites: FC<Props> = ({ pendingInvites }) => {
```

File: src/components/group/pending-invites.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+          disableSelectionOnClick
+          onRowClick={ () => {} }
+        />
+      </Paper>
+    );
+  };
+
+  return (
+    <>
+      { renderPendingInvitesDataTable() }
```
Comment (lauriLukkarinen) [inline src/components/group/pending-invites.tsx:63]: Same thing here, render the table directly in the component return.

File: src/components/group/pending-invites.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+    return (
+      <Paper>
+        <DataGrid
+          autoHeight
+          rows={ pendingInvites }
+          columns={ columns }
+          pageSize={ 10 }
+          disableSelectionOnClick
+          onRowClick={ () => {} }
```
Comment (lauriLukkarinen) [inline src/components/group/pending-invites.tsx:55]:
```suggestion
```

File: src/components/group/pending-invites.tsx [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        renderCell: ({ value, row }) => (
+          <Box>
+            <span>
+              { value }
+            </span>
+            <Button
+              variant="text"
+              endIcon={ <SendIcon/> }
+              // TODO: Resend invitation logic
+              onClick={ () => console.log(row)}
```
Comment (lauriLukkarinen) [inline src/components/group/pending-invites.tsx:37]:
```suggestion
              // TODO: Resend invitation logic
              // onClick={ ... }
```

File: src/components/group/pending-requests.tsx [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+              style={{ marginRight: 8 }}
+              // TODO: Reject request logic
+              onClick={ () => console.log(row)}
+            >
+              { strings.groupManagementScreen.pendingRequestsScreen.reject }
+            </Button>
+            <Button
+              color="primary"
+              // TODO: Accept request logic
+              onClick={ () => console.log(row)}
```
Comment (lauriLukkarinen) [inline src/components/group/pending-requests.tsx:52]:
```suggestion
              // TODO: Reject request logic
              // onClick={ ... }
            >
              { strings.groupManagementScreen.pendingRequestsScreen.reject }
            </Button>
            <Button
              color="primary"
              // TODO: Accept request logic
              // onClick={ ... }
```

File: src/components/group/pending-requests.tsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    return (
+      <Paper>
+        <DataGrid
+          autoHeight
+          rows={ pendingRequests }
+          columns={ columns }
+          pageSize={ 10 }
+          disableSelectionOnClick
+          onRowClick={ () => {} }
```
Comment (lauriLukkarinen) [inline src/components/group/pending-requests.tsx:69]:
```suggestion
```

File: src/components/group/pending-requests.tsx [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        />
+      </Paper>
+    );
+  };
+
+  return (
+    <>
+      { renderPendingRequestsDataTable() }
```
Comment (lauriLukkarinen) [inline src/components/group/pending-requests.tsx:77]: Same here with the rendering of the table

File: src/components/screens/group-dialog-screen.tsx [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+ */
+const GroupDialogScreen: FC = () => {
+  const dispatch = useAppDispatch();
+  const keycloak = useAppSelector(selectKeycloak);
+  const navigate = useNavigate();
+  const errorContext = useContext(ErrorContext);
+  const [usersGroups, setUsersGroups] = useState<UserGroup[]>([]);
+  const [allGroups, setAllGroups] = useState<UserGroup[]>([]);
+  const [ filteredGroups, setFilteredGroups ] = useState<UserGroup[]>([]);
+  const [loading, setLoading] = useState(false);
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:26]:
```suggestion
  const [ usersGroups, setUsersGroups ] = useState<UserGroup[]>([]);
  const [ allGroups, setAllGroups ] = useState<UserGroup[]>([]);
  const [ filteredGroups, setFilteredGroups ] = useState<UserGroup[]>([]);
  const [ loading, setLoading ] = useState(false);
```

File: src/components/screens/group-dialog-screen.tsx [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+        });
+
+        setDialogStatus(GroupDialogStatus.REQUEST_SENT);
+        setLoading(true);
+      } catch (error) {
+        errorContext.setError(strings.errorHandling.groupDialogsScreen.createGroupRequest);
+      }
+    } else {
+      errorContext.setError("Joining mutliple groups not yet supported");
+    }
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:212]: Reverse the condition:
```suggestion
    // TODO: currently set up to only join one group at a time
    if (groupsToJoin.length !== 1) {
      errorContext.setError("Joining multiple groups not yet supported");
      return;
    }

    if (!matchedGroup?.id) return;

    setLoading(true);

    try {
      await Api.getGroupJoinRequestsApi(keycloak.token).createGroupJoinRequest({
        groupId: matchedGroup.id,
        groupJoinRequest: {
          email: keycloak.tokenParsed.email,
          groupId: matchedGroup.id
        }
      });

      setDialogStatus(GroupDialogStatus.REQUEST_SENT);
      setLoading(true);
    } catch (error) {
       errorContext.setError(strings.errorHandling.groupDialogsScreen.createGroupRequest);
    }
```

File: src/components/screens/group-dialog-screen.tsx [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+  };
+
+  /**
+   * Renders the welcome dialog
+   */
+  const renderWelcomeDialog = () => (
+    <GenericDialog
+      error={ false }
+      open={ true }
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:221]:
```suggestion
      open
```

File: src/components/screens/group-dialog-screen.tsx [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+    </GenericDialog>
+  );
+
+  /**
+   * Renders the create dialog
+   */
+  const renderCreateDialog = () => (
+    <GenericDialog
+      error={ false }
+      open={ true }
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:242]:
```suggestion
      open
```

File: src/components/screens/group-dialog-screen.tsx [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+      onClose={ () => {} }
+      onCancel={ () => setDialogStatus(GroupDialogStatus.WELCOME) }
+      onConfirm={ createGroup }
+      title={ strings.groupDialogsScreen.createDialog.title }
+      positiveButtonText={ strings.groupDialogsScreen.createNewGroup }
+      cancelButtonText={ strings.groupDialogsScreen.cancel }
+      disabled={ !newGroupName }
+      hideClose={ true }
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:250]:
```suggestion
      hideClose
```

File: src/components/screens/group-dialog-screen.tsx [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+   * Renders the create done dialog
+   */
+  const renderCreateDoneDialog = () => (
+    <GenericDialog
+      open={ true }
+      onClose={ () => {} }
+      onConfirm={ () => navigate(`/groups/${usersGroups[0].id}`)}
+      title={ strings.groupDialogsScreen.createDoneDialog.title }
+      positiveButtonText={ strings.groupDialogsScreen.createDoneDialog.goToGroupManagement }
+      hideClose={ true }
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:274]:
```suggestion
      open
      onClose={ () => {} }
      onConfirm={ () => navigate(`/groups/${usersGroups[0].id}`)}
      title={ strings.groupDialogsScreen.createDoneDialog.title }
      positiveButtonText={ strings.groupDialogsScreen.createDoneDialog.goToGroupManagement }
      hideClose
```

File: src/components/screens/group-dialog-screen.tsx [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+        error={ false }
+        open={ true }
+        onClose={ () => {} }
+        onCancel={ () => setDialogStatus(GroupDialogStatus.WELCOME) }
+        onConfirm={ joinGroup }
+        title={ strings.groupDialogsScreen.joinDialog.title }
+        positiveButtonText={ strings.groupDialogsScreen.joinDialog.sendRequest }
+        cancelButtonText={ strings.groupDialogsScreen.createNewGroup }
+        disabled={ !isChecked }
+        hideClose={ true }
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:299]:
```suggestion
        open
        onClose={ () => {} }
        onCancel={ () => setDialogStatus(GroupDialogStatus.WELCOME) }
        onConfirm={ joinGroup }
        title={ strings.groupDialogsScreen.joinDialog.title }
        positiveButtonText={ strings.groupDialogsScreen.joinDialog.sendRequest }
        cancelButtonText={ strings.groupDialogsScreen.createNewGroup }
        disabled={ !isChecked }
        hideClose
```

File: src/components/screens/group-dialog-screen.tsx [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+        { strings.groupDialogsScreen.createDoneDialog.content }
+      </Typography>
+    </GenericDialog>
+  );
+
+  /**
+   * Render the join group dialog
+   */
+  const renderJoinDialog = () => {
+    const isChecked = joinGroupCheckedItems.some(checked => checked);
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:286]: I believe this is the correct comparison?
```suggestion
    const isChecked = joinGroupCheckedItems.some(item => item.checked);
```

File: src/components/screens/group-dialog-screen.tsx [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+      open={ true }
+      onClose={ () => {} }
+      onCancel={ () => setDialogStatus(GroupDialogStatus.WELCOME) }
+      onConfirm={ () => {} }
+      title={ strings.groupDialogsScreen.requestSentDialog.title }
+      positiveButtonText={ strings.groupDialogsScreen.requestSentDialog.requestSent }
+      cancelButtonText={ strings.groupDialogsScreen.createNewGroup }
+      disabled={ true }
+      hasIcon={ true }
+      hideClose={ true }
```
Comment (lauriLukkarinen) [inline src/components/screens/group-dialog-screen.tsx:344]:
```suggestion
      open
      onClose={ () => {} }
      onCancel={ () => setDialogStatus(GroupDialogStatus.WELCOME) }
      onConfirm={ () => {} }
      title={ strings.groupDialogsScreen.requestSentDialog.title }
      positiveButtonText={ strings.groupDialogsScreen.requestSentDialog.requestSent }
      cancelButtonText={ strings.groupDialogsScreen.createNewGroup }
      disabled
      hasIcon
      hideClose
```

File: src/components/screens/group-management-routes.tsx [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+  groupId?: string;
+  groupMembers: User[];
+  pendingRequests: GroupJoinRequest[];
+  pendingInvites: GroupJoinInvite[];
+}
+
+/**
+ * Component for group routes
+ */
+const GroupRoutes: React.FC<Props> = ({ groupId, groupMembers, pendingRequests, pendingInvites }) => {
```
Comment (lauriLukkarinen) [inline src/components/screens/group-management-routes.tsx:21]:
```suggestion
/**
 * Component for group routes
 *
 * @param props component properties
 */
const GroupRoutes: React.FC<Props> = ({ groupId, groupMembers, pendingRequests, pendingInvites }) => {
```

File: src/components/screens/group-management-screen.tsx [hunk 23]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    setLoading(true);
+    try {
+      const foundPendingRequests = await Api.getGroupJoinRequestsApi(keycloak.token).listGroupJoinRequests({
+        groupId: groupId,
+        status: JoinRequestStatus.Pending
+      });
+
+      console.log("pending requests", foundPendingRequests);
```
Comment (lauriLukkarinen) [inline src/components/screens/group-management-screen.tsx:98]:
```suggestion
```

File: src/components/screens/group-management-screen.tsx [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    setLoading(true);
+    try {
+      const allInvites = await Api.getGroupJoinInvitesApi(keycloak.token).listGroupJoinInvites({
+        groupId: groupId
+      });
+
+      const foundPendingInvites = allInvites.filter(invite => invite.status === JoinRequestStatus.Pending);
+      setPendingInvites(foundPendingInvites);
```
Comment (lauriLukkarinen) [inline src/components/screens/group-management-screen.tsx:123]:
```suggestion
      setPendingInvites(allInvites.filter(invite => invite.status === JoinRequestStatus.Pending));
```

File: src/components/screens/group-management-screen.tsx [hunk 25]
```diff
@@ -0,0 +1,299 @@
...
+          groupId: groupId,
+          status: JoinRequestStatus.Pending
+        }
+      });
+
+      setLoading(false);
+      setInviteDialogOpen(false);
+    } catch (error) {
+      errorContext.setError(strings.errorHandling.groupManagementScreen.inviteMember);
+    }
```
Comment (lauriLukkarinen) [inline src/components/screens/group-management-screen.tsx:191]:
```suggestion
    try {
      await Api.getGroupJoinInvitesApi(keycloak.token).createGroupJoinInvite({
        groupId: groupId,
        groupJoinInvite: {
          email: inviteMemberEmail,
          groupId: groupId,
          status: JoinRequestStatus.Pending
        }
      });

      setInviteDialogOpen(false);
    } catch (error) {
      errorContext.setError(strings.errorHandling.groupManagementScreen.inviteMember);
    }

    setLoading(false);
```

File: src/components/screens/group-management-screen.tsx [hunk 26]
```diff
@@ -0,0 +1,299 @@
...
+        <CircularProgress color="inherit" size={ 60 }/>
+      </Box>
+    );
+  };
+
+  return (
+    <>
+      { renderInviteDialog() }
+      <SidePanelLayout
+      // TODO: Missing subtitle
```
Comment (lauriLukkarinen) [inline src/components/screens/group-management-screen.tsx:278]:
```suggestion
        // TODO: Missing subtitle
```

File: src/localization/fi.json [hunk 27]
```diff
@@ -561,5 +572,58 @@
     "title": "Sinun tulee kirjautua sisään sivustolle, jolle olet luomassa ilmoitusta",
     "registerText": "Eikö sinulla ole tiliä valitulla sivulla?",
     "registerLink": "Luo tili sivulle"
+  },
+  "groupDialogsScreen": {
+    "createNewGroup": "Luo uusi ryhmä" ,
+    "cancel": "Peruuta",
+    "welcomeDialog": {
+      "title": "Tervetuloa purkukartoitus.fi palveluun!",
+      "joinGroup": "Liity Rhymään" ,
```
Comment (lauriLukkarinen) [inline src/localization/fi.json:581]:
```suggestion
      "joinGroup": "Liity Ryhmään" ,
```

File: src/localization/fi.json [hunk 28]
```diff
@@ -561,5 +572,58 @@
...
+    "joinDialog": {
+      "title": "Liity ryhmään",
+      "sendRequest": "Lähetä liittymispyyntö",
+      "content": "Pystyäksesi tekemään kartoituksia on sinun kuuluttava johonkin ryhmään. Hae organisaatiosi ryhmää alla olevasta listasta. Jos organisaatiollasi ei ole ryhmää, voit luoda uuden ryhmän valitsemalla 'Luo uusi ryhmä'.",
+      "placeholder": "Hae ryhmää"
+    },
+    "requestSentDialog": {
+      "title": "Liity ryhmään",
+      "requestSent": "Pyyntö lähetetty",
+      "content": "Liittymispyyntö lähetetty ryhmään XAMK. Saat ilmoituksen sähköpostiisi, kun pyyntö on hyväksytty. \n Hyväksynnän jälkeen pääset tarkastelemaan organisaatiosi kartoituksia ja luomaan uusia."
```
Comment (lauriLukkarinen) [inline src/localization/fi.json:603]:
```suggestion
      "content": "Liittymispyyntö lähetetty ryhmään {0}. Saat ilmoituksen sähköpostiisi, kun pyyntö on hyväksytty. \n Hyväksynnän jälkeen pääset tarkastelemaan organisaatiosi kartoituksia ja luomaan uusia."
```

File: src/localization/fi.json [hunk 29]
```diff
@@ -561,5 +572,58 @@
...
+    }
+  },
+  "groupManagementScreen": {
+    "groupMembers": "Ryhmän jäsenet",
+    "pendingInvites": "Lähetetyt kutsut ({0})",
+    "pendingRequests": "Jäsenpyynnöt ({0})",
+    "inviteUser": "Kutsu käyttäjä",
+    "sendInvite": "Kutsu käyttäjä",
```
Comment (lauriLukkarinen) [inline src/localization/fi.json:611]:
```suggestion
    "sendInvite": "Lähetä kutsu",
```

File: src/types/index.ts [hunk 30]
```diff
@@ -105,10 +105,27 @@ export interface SurveySummary {
   attachments: Attachment[];
 }
 
+export interface CheckboxData {
```
Comment (lauriLukkarinen) [inline src/types/index.ts:108]:
```suggestion
/**
 * Checkbox state data
 */
export interface CheckboxData {
```

**Review by lauriLukkarinen — CHANGES_REQUESTED** [review]

File: src/components/group/group-members.tsx (hunk 5 above)
Comment (Jdallos) [reply src/components/group/group-members.tsx:91]: Done, with useMemo also

File: src/components/group/pending-invites.tsx (hunk 7 above)
Comment (Jdallos) [reply src/components/group/pending-invites.tsx:64]: done

File: src/localization/fi.json [hunk 31]
```diff
@@ -152,13 +152,29 @@
       "loginFailed": "Väärä käyttäjätunnus tai salasana",
       "serverError": "Valitun sivun kanssa on yhteysongelmia. Yritä myöhemmin uudelleen."
     },
+    "groupDialogsScreen": {
+      "listGroups": "Virhe haettaessa ryhmiä",
+      "createGroup": "Virhe luotaessa ryhmää",
+      "createGroupRequest": "Virhe luotaessa ryhmään liittymispyyntöä",
```
Comment (lauriLukkarinen) [inline src/localization/fi.json:158]:
```suggestion
      "createGroupRequest": "Virhe luotaessa liittymispyyntöä ryhmään",
```

**Review by lauriLukkarinen — APPROVED** [review]

File: src/components/screens/group-dialog-screen.tsx [hunk 32]
```diff
@@ -0,0 +1,551 @@
...
+    // TODO: currently set so user can only join a single group
+    const matchedGroup = allGroups.find(group => group.name.toLowerCase() === groupsToJoin[0].name.toLowerCase());
+
+    if (!matchedGroup || !keycloak?.token || !keycloak?.tokenParsed) return;
+
+    setJoinedGroup(matchedGroup?.name);
+
+    // TODO: currently set up to only join one group at a time
+    if (groupsToJoin.length !== 1) {
+      errorContext.setError("Joining multiple groups not yet supported");
```
[the remark this answers is not in the document]
Comment (EVP1989) [reply src/components/screens/group-dialog-screen.tsx]: Fixed

**Pull request merged**